internal/common/customtype/object.go

```diff
@@ -0,0 +1,209 @@
+	attrValue, err := t.ObjectType.ValueFromTerraform(ctx, in)
+
+	if err != nil {
+		return nil, err
+	}
+
+	objectValue, ok := attrValue.(basetypes.ObjectValue)
+	if !ok {
+		return nil, fmt.Errorf("unexpected value type of %T", attrValue)
+	}
+
+	objectValuable, diags := t.ValueFromObject(ctx, objectValue)
+	if diags.HasError() {
+		return nil, fmt.Errorf("unexpected error converting ObjectValue to ObjectValuable: %v", diags)
+	}
+
+	return objectValuable, nil
+}
+
+func (t ObjectType[T]) ValueType(_ context.Context) attr.Value {
+	return ObjectValue[T]{}
+}
+
+type ObjectValue[T any] struct {
+	basetypes.ObjectValue
+}
+
+type ObjectValueInterface interface {
+	basetypes.ObjectValuable
+	ValuePtrAsAny(ctx context.Context) (any, diag.Diagnostics)
+	NewObjectValue(ctx context.Context, value any) ObjectValueInterface
+	NewObjectValueNull(ctx context.Context) ObjectValueInterface
+}
+
+func (v ObjectValue[T]) NewObjectValue(ctx context.Context, value any) ObjectValueInterface {
+	attrTypes, diags := getAttributeTypes[T](ctx)
+	if diags.HasError() {
+		panic(fmt.Errorf("error creating ObjectValue: %v", diags))
+	}
+
+	newValue, diags := basetypes.NewObjectValueFrom(ctx, attrTypes, value)
+	if diags.HasError() {
+		return newObjectValueUnknown[T](ctx)
+	}
+
+	return ObjectValue[T]{ObjectValue: newValue}
+}
+
+func (v ObjectValue[T]) NewObjectValueNull(ctx context.Context) ObjectValueInterface {
+	return newObjectValueNull[T](ctx)
+}
+
+func newObjectValueNull[T any](ctx context.Context) ObjectValue[T] {
+	attrTypes, diags := getAttributeTypes[T](ctx)
+	if diags.HasError() {
+		panic(fmt.Errorf("error creating null ObjectValue: %v", diags))
+	}
+	return ObjectValue[T]{ObjectValue: basetypes.NewObjectNull(attrTypes)}
+}
+
+func newObjectValueUnknown[T any](ctx context.Context) ObjectValue[T] {
+	attrTypes, diags := getAttributeTypes[T](ctx)
+	if diags.HasError() {
+		panic(fmt.Errorf("error creating unknown ObjectValue: %v", diags))
+	}
+	return ObjectValue[T]{ObjectValue: basetypes.NewObjectUnknown(attrTypes)}
+}
+
+func (v ObjectValue[T]) Equal(o attr.Value) bool {
+	other, ok := o.(ObjectValue[T])
+	if !ok {
+		return false
+	}
+	return v.ObjectValue.Equal(other.ObjectValue)
+}
+
+func (v ObjectValue[T]) Type(ctx context.Context) attr.Type {
+	return NewObjectType[T](ctx)
+}
+
+func (v ObjectValue[T]) ValuePtrAsAny(ctx context.Context) (any, diag.Diagnostics) {
+	valuePtr := new(T)
+
+	if v.IsNull() || v.IsUnknown() {
+		return valuePtr, nil
+	}
+
+	diags := v.As(ctx, valuePtr, basetypes.ObjectAsOptions{})
+	if diags.HasError() {
+		return nil, diags
+	}
+
+	return valuePtr, diags
+}
+
+func getAttributeTypes[T any](ctx context.Context) (map[string]attr.Type, diag.Diagnostics) {
+	var t T
+	return valueToAttributeTypes(ctx, reflect.ValueOf(t))
+}
+
+func valueToAttributeTypes(ctx context.Context, value reflect.Value) (map[string]attr.Type, diag.Diagnostics) {
```

Comment:
q: Will there be potential for code reuse when implementing custom type of other nested types?